File: Data Structures & Algorithms/kth-smallest-integer-in-bst/submission-4.js
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     constructor(val = 0, left = null, right = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    /**
     * @param {TreeNode} root
     * @param {number} k
     * @return {number}
     */
    kthSmallest(root, k) {
        const stack = [];
        stack.push(root);

        let current = root;

        while (current != null || stack.length > 0) {
            while (current != null) {
                stack.push(current);
                current = current.left;
            }
            current = stack.pop();
            k--; 
            if (k === 0) return current.val;
            current = current.right;
        }
        
    }
}

File: Data Structures & Algorithms/binary-tree-diameter/submission-4.py
class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        self.best = 0
        self.height(root)
        return self.best

    def height(self, node: Optional[TreeNode]) -> int:
        if not node:
            return 0

        left = self.height(node.left)
        right = self.height(node.right)

        if left + right > self.best:
            self.best = left + right

        if left > right:
            return left + 1
        return right + 1